npeps/npep-122.md

```diff
@@ -158,7 +158,344 @@ Fourth, the ANP subject allows using pod selectors, while tenancy use cases only
+
+```yaml
+kind: AdminNetworkPolicy
+spec:
+  priority: 10
+  subject:
+    namespaces:
+      matchLabels:
+        kubernetes.io/metadata.name: sensitive-ns
+  ingress:
+    - action: Allow
+      from:
+      - tenant:
+          labels: ["user"]
+          tenant: Same
+```
+
+**CONS** complicates selection or affected namespaces, allows strange configuration, where namespaces selected by 
+`subject` have no tenancy labels. Has the same problem as the original design.
+
+3. Create 2 objects with ANP and BANP priorities (let's say NetworkTenancy and BaselineNetworkTenancy)
+
+```yaml
+kind: NetworkTenancy
+spec:
+  priority: 10
+  tenancyLabels: ["user"]
+  actions: Allow
+  peer: SameTenant
+---
+kind: BaselineNetworkTenancy
+spec:
+  priority: 10
+  tenancyLabels: ["user"]
+  actions: Allow
+  peer: SameTenant
+```
+
+While multiple ANPs with the same priority are allowed, we probably can allow multiple Tenancies or Tenancy and ANP
+with the same priority, but if we decide to only allow ANP per priority, Tenancy needs to be accounted for in the same range.
+
+**CONS**: BANP doesn't have a priority, to use this method we would need to define a priority for BANP.
+
+4. Create 1 new object with implicit priorities.
+
+`precedence` field + reserved highest-priority rule before (B)ANP
+Similar to the previous one, but a bit more flexible:
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels: ["user"]
+  precedence: ANP/BANP
+  action: Allow/Skip
+```
+**CONS** Priorities are implicit and need to be added as extra layers between ANP/NP/BANP.
+**PROS** 
+- No changes to the existing ANP/BANP objects
+- Limited to the use cases we designed it for (smaller chance to shoot yourself in a foot)
+- Users that don't care about tenancy can just ignore this CRD
+- We can throw it away if we want to change API again
+
+
+#### Final suggestion
+
+Considering we agree with option 4 from the previous section on priorities specification, we can outline further details here.
+
+For
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels: ["user"]
+  precedence: ANP/BANP
+  action: Allow/Skip
+```
+there are only 3 useful combinations of precedence+action (BANP+Skip is the same as BANP+Allow), so we can express these
+options in a different way, e.g. define 3 modes
+- always-allow (ANP+Allow)
+- delegate-to-NP (ANP+Skip)
+- baseline-allow (BANP+Allow)
+
+<details>
+<summary>Full yaml examples (with the initial fields, will be updates as we agree on the final CRD format)</summary>
+
+* 4.1 "overridable isolation"
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: BANP
+  action: Allow
+---
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
```

Comment:
first of all, I am not sure between-tenant behaviour is a necessary part of the tenancy. It is not obvious why "other tenants" should be different "other namespaces". Not sure if denying traffic for a namespace outside your tenant should be based on whether that namespace is also a part of some other tenant or not. 
Not saying it shouldn't be there, but the core concept of tenancy is `sameTenant`, which is also a dynamic selector that is not possible to express in any other way. Performance-wise I would probably even expect `notSameTenant` to be represented as a lower-priority deny rule for tenancy selector (with higher-priority `sameTenant` rule), as you don't need to create dynamic `notSameTenant` sets of pods for every tenant, but can only do that for `sameTenant`.
About including full use case (both allow and deny with their priorities): the main problem I faced is integration with ANP/BANP. When you only use `sameTenant,` you need only 1 implicit priority (for every use case we have highest-ANP and highest-BANP would work). But with deny rules, you may want deny as also highest-priority (use case 4.2) or lower than allow-from-monitoring (use case 4.4.1). Considering implicit priorities are adding more complexity, trying to track where exactly each one fits both for allow and deny rules just complicates the design (e.g. I am not sure how https://github.com/kubernetes-sigs/network-policy-api/pull/178#discussion_r1504764041 this example defined tenancy priority relative to ANP priority)

Reply:
> first of all, I am not sure between-tenant behaviour is a necessary part of the tenancy.

Wow. OK, I guess I understand why we're having trouble agreeing on the details of the API now.

As I see it, tenancy is 100% about between-tenant behavior and 0% about anything else.

Look at the original user story:

> As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default. Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant may own more than 1 Namespace.

It only talks about isolation between tenants, not anything that happens _inside_ the tenants. And the updated 4.1 says

> ... By default, cross-tenant traffic is dropped. ...

and again, neither 4.1 nor 4.2 says anything about what happens inside tenants.

The goal of tenancy is to put up walls between certain sets of namespaces. That's it. Traffic between tenants is restricted in a particular way, but all traffic that isn't between-tenant behaves exactly like it would have without the tenant definitions; you can make rules like "everybody can talk to the apiserver" or "namespaces with a user label can't talk to the internet by default", but you don't need a tenancy API to do that. You only need the tenancy API to specify between-tenant behavior.

Reply:
Not sure if it helps, coming here from end-user/implementor perspective. The one single ask I have been getting always is: putting wall up between tenants (tenant here being set of namespaces); the users I know are designing almost always: "allow to youself, (isSame) deny to all other tenants(isNotSame)" => I know this is covered in one of the user stories already but re-hashing that that is the single most thing the current ANP/BANP can't do well enough; I think other things like allow {} to dns and deny {} can be done using anp/banp rules unless I am missing something.

- prio1 allow to your own tenant (I think this is 4.3 but ANP level? I have not yet heard a ask for BANP level)
- prio2 allow to DNS
- prio3 allow from monitoring
- last prio deny everything to everything total lockdown

the prio1 rule expression is what we struggle with on ANP because putting {} in the subject doesn't work as it does not allow reflecting tenancy in peers, rest can be done using anp/banp rules. so really the ask is almost always an easy way to say "please allow to my own tenant" because the deny is almost always going to be the default deny banp at the bottom which takes care of putting up walls. OR if you do the {} to {} deny in ANP that also solves that problem (I can see how the deny to other tenants can be useful if we can express that explicitly)